acceptancetests/jujupy/k8s_provider/gke.py

```diff
@@ -0,0 +1,283 @@
+from jujupy.utility import until_timeout
+
+from .base import (
+    Base,
+    K8sProviderType,
+)
+from .factory import register_provider
+
+
+logger = logging.getLogger(__name__)
+CLUSTER_STATUS = container_v1.enums.Cluster.Status
+
+
+@register_provider
+class GKE(Base):
+
+    name = K8sProviderType.GKE
+
+    driver = None
+    gke_cluster_name = None
+    gke_cluster = None
+
+    default_params = None
+
+    def __init__(self, bs_manager, timeout=1800):
+        super().__init__(bs_manager, timeout)
+
+        self.gke_cluster_name = self.client.env.controller.name  # use controller name as cluster name
+        self.default_storage_class_name = ''
+        self.__init_driver(bs_manager.client.env)
+
+    def __init_driver(self, env):
+        zone = env.get_host_cloud_region()[2] + '-b'
+        cfg = env._config
+        self.default_params = dict(
+            project_id=cfg['project_id'],
+            zone=zone,
+        )
+
+        self.driver = container_v1.ClusterManagerClient(
+            credentials=service_account.Credentials.from_service_account_info(cfg)
+        )
+        
+        self.__ensure_gcloud(cfg)
+
+        # list all running clusters
+        running_clusters = self.driver.list_clusters(**self.default_params)
+        logger.info('running gke clusters: %s', running_clusters)
+
+    def _ensure_cluster_stack(self):
+        self.provision_gke()
+        
+    def __ensure_gcloud(self, cfg):
+        gcloud_bin = shutil.which('gcloud')
+        if gcloud_bin is None:
+            # ensure gcloud installed.
+            self.sh('sudo', 'snap', 'install', 'google-cloud-sdk', '--classic')
+            logger.debug("gcloud installed successfully")
+            gcloud_bin = shutil.which('gcloud')
+        
+        cred = {
+            k: cfg[k] for k in (
+                'project_id', 
+                'private_key_id', 
+                'private_key', 
+                'client_email', 
+                'client_id', 
+                'auth_uri', 
+                'token_uri', 
+                'auth_provider_x509_cert_url', 
+                'client_x509_cert_url',
+            )
+        }
+        cred['type'] = 'service_account'
+        cred_file = tempfile.NamedTemporaryFile(suffix='.json', delete=False)
+        try:
+            with open(cred_file.name, 'w') as f:
+                json.dump(cred, f)
+                
+            # set credential;
+            o = self.sh(gcloud_bin, 'auth', 'activate-service-account', '--key-file', cred_file.name)
+            logger.debug(o)
+            # set project;
+            o = self.sh(gcloud_bin, 'config', 'set', 'project', cfg['project_id'])
+            logger.debug(o)
+        finally:
+            os.unlink(cred_file.name)
+
+    def _tear_down_substrate(self):
+        try:
+            self.driver.delete_cluster(
+                cluster_id=self.gke_cluster_name, **self.default_params,
+            )
+        except exceptions.NotFound:
+            ...
```

Comment:
It might just be that I haven't done any Python for ages, but this seems like weird style to me. I can see using it for abstract methods where you don't want any implementation, but in normal blocks I think `pass` would be better. (I may be out of touch on this one.)

Reply:
sure, changed to pass.